Openshift backup directories, databases and restore them. (#814)
UKCloud/documentation: articles/openshift/oshift-how-backup-directories-database.md

---
title: How to backup container directories, databases and restore them on OpenShift | UKCloud Ltd
description: Provides guidance on how customers can backup container directories, databases and restore them. 
services: openshift
author: Daniel Brennand
reviewer: 
lastreviewed: 

toc_rootlink: How To
toc_sub1: 
toc_sub2:
toc_sub3:
toc_sub4:
toc_title: Backup container directories, databases and restore them.
toc_fullpath: How To/oshift-how-backup-directories-database.md
toc_mdlink: oshift-how-backup-directories-database.md
---

# How to backup container directories, databases and restore them

## Overview

> This article will explain how you can backup container directories, databases (specifically [PostgreSQL](https://www.postgresql.org/)) to your localhost. Furthermore, it will also go through how you can restore them.

Reviewer: Using "localhost" here feels like an odd turn of phrase.
```suggestion
This article will explain how you can backup container directories, databases (specifically [PostgreSQL](https://www.postgresql.org/)) to your local client. Furthermore, it will also go through how you can restore them.
```

Reviewer 2: I prefer local client as well, if that counts for anything ;)

### Prerequisites

This guide assumes that you have a UKCloud for OpenShift envrionment and can access it using the [OpenShift Container Platform command line interface](https://docs.openshift.com/container-platform/3.11/cli_reference/index.html#cli-reference-index).

Two OC CLI commands will be utilised in this article:

1. [`oc exec`](https://docs.openshift.com/container-platform/3.11/dev_guide/executing_remote_commands.html)
2. [`oc rsync`](https://docs.openshift.com/container-platform/3.11/dev_guide/copy_files_to_container.html#overview)

Begin by logging into your UKCloud for OpenShift envrionment using the following command:

```bash
oc login --server {server} --token {token here}
```

> [!NOTE]
> The `--token` can be retrieved by logging into the OpenShift web interface, 
> clicking your username in the top right and clicking **Copy Login Command**.

> ## Backing up container directories to your localhost

Reviewer: As above 
```suggestion
## Backing up container directories to your local client
```

> 1. Create a directory on your localhost for files to be backed up to:

Reviewer: ```suggestion
1. Create a directory on your local client for files to be backed up to:
```

## Upload a directory to a container

1. Switch to the project which the container you want to upload a directory too is located:

    ```bash
    oc project {project name}
    ```

    > [!NOTE]
    > To list all projects use: `oc projects`.

2. Identify the pod which the container you want to upload a directory to resides in:

    ```bash
    oc get pods
    ```

4. Using `oc rsync`, upload the directory to the container:

    ```bash
    # oc rsync usage:
    oc rsync {destination/on/host} {podname}:{path/to/upload}

    # Example usage:
    oc rsync ~/oshiftbackups/files mypod-1-hscwm:/tmp/files
    ```

    > [!NOTE]
    > If you have multiple containers in a pod, use the `-c {container name}` parameter to specify a specific container name. <br>
    > Example: `oc rsync {podname}:{path/to/directory} {destination/on/host} -c {container name}`

## Backing up a database on OpenShift

In this example, the type of database which will be backed up is PostgreSQL. Substitute the commands executed in this guide (using `oc exec`) with commands which are appropriate to the database type you're using.

> 1. Create a directory on your localhost for files to be backed up to:

Reviewer: ```suggestion
1. Create a directory on your local client for files to be backed up to:
```

    ```bash
    mkdir ~/oshiftdatabasebackup
    ```

2. Switch to the project where your database pod is located:

    ```bash
    oc project {project name}
    ```

    > [!NOTE]
    > To list all projects use: `oc projects`.

3. Identify the pod which the database container you want to backup resides in:

    ```bash
    oc get pods
    
    NAME                           READY    STATUS      RESTARTS    AGE
    postgresql-1-hscwm   1/1           Running    0                   31m
 
    # In this example, the database is running in pod: postgresql-1-hscwm
    ```

4. Using `oc exec`, run the following command to execute the PostgreSQL backup command in the container:

    ```bash
    # Execute a remote command on a container
    # Dump the contents of the database and pipe it into gzip (to create a compressed archive)
    # oc exec usage:
    oc exec {podname} -- bash -c 'pg_dump {databasename} | gzip > /opt/app-root/src/{databasename}.gz'
    
    # Example usage:
    oc exec postgresql-1-hscwm -- bash -c 'pg_dump sampledb | gzip > /opt/app-root/src/sampledb.gz'
    ```

> 5. Using `oc rsync`, copy the database backup file contaning the PostgreSQL database files to your localhost:

Reviewer: ```suggestion
5. Using `oc rsync`, copy the database backup file contaning the PostgreSQL database files to your local client:
```

    ```bash
    # oc rsync usage:
    oc rsync {podname}:{path/to/directory} {destination/on/host}
 
    # Example usage:
    # This command copies the database backup file to /home/databasebackup
    oc rsync postgresql-1-hscwm:/opt/app-root/src/sampledb.gz ~/oshiftdatabasebackup
    ```

    > [!NOTE]
    > If you have multiple containers in a pod, use the `-c {container name}` parameter to specify a specific container name. <br>
    > Example: `oc rsync {podname}:{path/to/directory} {destination/on/host} -c {container name}`

> Your database is now backed up onto your localhost at `/home/oshiftdatabasebackup`.

Reviewer: ```suggestion
Your database is now backed up onto your local client in `/home/oshiftdatabasebackup`.
```
